### PR TITLE
HTTPRoute: allow more match clauses

### PR DESCRIPTION
**Pull request by howardjohn**
Today, a route is limited to 16 routes with 8 matches each. This is problematic in real world environments.

While its easy to split a route up (you can easily get the same behavior, and similar cognitive complexity by having 1 route per HTTPRoute), its quite hard to split up matches. For instance, https://github.com/istio/istio/blob/d7a9700d5eaa4e9728274b408623670f48deadb5/samples/bookinfo/gateway-api/bookinfo-gateway.yaml#L23-L38 is an example of a route matching the exposed APIs for a frontend. This is only a small tivial example, and it already uses 5. In our user base, we often see far beyond 8.

Splitting these up is both complex for the user, and may actually lead to different behaviors if implementations treat route groups differently (for instance, if we add retry budget -- that could now be split; other core or implementation specific policies may behave similarly, this is just an example).

The limit on 8 here seems quite small given the context of real usage, the other limits in the API (16 routes, 64 listeners), and the cost of manually working around it. Based on this, I propose we raise the limit to match Gateway listeners.

**What type of PR is this?**

/kind feature

**Does this PR introduce a user-facing change?**:

```release-note
Increased the limit on HTTPRoute matches from 8 to 64.
```

### CODE REVIEW

**Comment by robscott** [comment]
@howardjohn Historically we've been against this because of the worst case scenarios this could lead to if you max out the length of every list within a list (16 rules * 8 matchers * 16 filters == 2048 possible filters in a Route). Increasing the matchers from 8 to 64 would turn that into 16384, which would obviously be problematic.

With that said, we've got CEL now, so we could likely make a case for the following:

1. Allow up to 64 matchers in a Rule (that seems rather extreme, but something >8 does seem sensible)
2. Write CEL validation that ensures that the maximum number of matchers allowed in a Route does not go up from it's current theoretical max (128)

That would mean we're allowing significantly more flexibility here without actually increasing the max size of an individual route. If this ends up working reasonably well, we could apply the same principle elsewhere in the API.

**Comment by howardjohn** [comment]
For my understanding, what specifically are we trying to constrain with the overall size limit? Is it...
* Not requiring implementations to support large routes
* Staying below CEL limits
* Making sure we fail hard before etcd size limits
* Something else?

If its the CEL limits, is the concern that its within the complexity limits now (which already take into account the 16*64), or that it could in the future if matches get more complex?

I ask since it may impact the solution here

**Comment by robscott** [comment]
> For my understanding, what specifically are we trying to constrain with the overall size limit? Is it...
> 
> * Not requiring implementations to support large routes
> * Staying below CEL limits
> * Making sure we fail hard before etcd size limits

All of the above. I'd also mention that it's possible that we'll add more types of matchers inside this list which will only increase the overall size and complexity, I don't want to get so close to a max size that we end up having no additional room to grow. I'm also strongly biased towards more smaller routes than a few very large routes. In general though, when we're talking about API compatibility, this is already a GA API, and significantly loosening the validation would be problematic. I think we can make a reasonably strong case in favor of a change that allows for more flexibility but still fits within the original constraints though.

**Comment by arkodg** [comment]
+1 to this change, this will reduce the friction users are facing when migrating from exiting APIs to Gateway API, who are used to authoring rules a specific way, and this is currently slowing down Gateway API adoption.
@robscott's suggestion of adding a CEL validation to measure and limit total filters for a Route, is a great idea to account for the concerns of the size of the final object persisting in the API server

crosss linking previous issues raised by users in the past
* https://github.com/kubernetes-sigs/gateway-api/issues/1485
* https://github.com/kubernetes-sigs/gateway-api/issues/2501
* https://github.com/envoyproxy/gateway/issues/396

**Comment by howardjohn** [comment]
I've pushed up a change to limit the aggregate size. LMK what you think

File: apis/v1/httproute_types.go
Comment (robscott) [inline apis/v1/httproute_types.go:122]: This is still a 2x increase over the current upper limits, would prefer to keep the upper limit at 128 unless we have a very clear need for more within a single route. Would also be great to have some basic test cases to ensure this validation is working as intended.

**Review by robscott — COMMENTED** [review]
Thanks @howardjohn!

File: apis/v1/httproute_types.go
Comment (howardjohn) [reply apis/v1/httproute_types.go]: Ok, dropped to 128 and added a positive and negative test

**Comment by robscott** [comment]
Sadly looks like we're running into https://github.com/kubernetes/kubernetes/issues/120973 - CRD is invalid because rule is too complex

**Comment by howardjohn** [comment]
That's only from the extra restriction fwiw. It works fine when we just allow it to be unbounded, even with the theoretical 16k entries.

**Comment by howardjohn** [comment]
Honestly don't get how that rule can possibly be 100x over the limit..

File: apis/v1/httproute_types.go
Comment (robscott) [inline apis/v1/httproute_types.go:122]: This is definitely gross, but the least bad we can offer until CEL cost estimation allows us to use `map` here (Kubernetes 1.30+). Here's the [CEL playground link](https://playcel.undistro.io/?content=H4sIAAAAAAAAA43TzQqCQBDA8VcZ9qS0ieNH6hJ17hnUg%2BhGgVqgXYrevV1qNugDhrnN8ltY%2FuxNtLoXSnjepPt9MB2v2vNhAyFswW7KsA6GZm4PeqJDBaEPi2r8EEgCuSIiEXFFTCLmioREwhUpiZQrViRWXJGRyLgiJ5FzRUGiYBd00ZFf%2FZ2d3R1deGSXR5ce2e3RxUd2fXT58Xd%2FH9bmAbmQomvmZjeeL7P5PlaoalzCSygoUUY1fK0kPgfln7HGXD6cOm3utZ%2Fz%2FgC1SMJupAMAAA%3D%3D) for future readers.

**Review by robscott — COMMENTED** [review]
Thanks @howardjohn! Will defer to someone else for LGTM.

/approve

**Comment by robscott** [comment]
Actually, would really like @youngnick specifically to sign off on this, will add a hold until he's able to take a look.

/hold

**Comment by gauravkghildiyal** [comment]
LGTM

Deferring review to Nick as per https://github.com/kubernetes-sigs/gateway-api/pull/3205#issuecomment-2243557250

File: apis/v1/httproute_types.go
Comment (sunjayBhatia) [inline apis/v1/httproute_types.go:122]:
```suggestion
	// +kubebuilder:validation:XValidation:message="While 16 rules and 64 matches per rule are allowed, the total matches must be less than 128",rule="(self.size() > 0 ? self[0].matches.size() : 0) + (self.size() > 1 ? self[1].matches.size() : 0) + (self.size() > 2 ? self[2].matches.size() : 0) + (self.size() > 3 ? self[3].matches.size() : 0) + (self.size() > 4 ? self[4].matches.size() : 0) + (self.size() > 5 ? self[5].matches.size() : 0) + (self.size() > 6 ? self[6].matches.size() : 0) + (self.size() > 7 ? self[7].matches.size() : 0) + (self.size() > 8 ? self[8].matches.size() : 0) + (self.size() > 9 ? self[9].matches.size() : 0) + (self.size() > 10 ? self[10].matches.size() : 0) + (self.size() > 11 ? self[11].matches.size() : 0) + (self.size() > 12 ? self[12].matches.size() : 0) + (self.size() > 13 ? self[13].matches.size() : 0) + (self.size() > 14 ? self[14].matches.size() : 0) + (self.size() > 15 ? self[15].matches.size() : 0) <= 128"
```

**Review by sunjayBhatia — COMMENTED** [review]
should we do the same for GRPCRoute as well while we're making this change?

**Comment by robscott** [comment]
> should we do the same for GRPCRoute as well while we're making this change?

I think updating GRPCRoute to match would be great. Would approve in this PR or a follow up.

**Comment by howardjohn** [comment]
Added gRPC route as well

**Comment by robscott** [comment]
Thanks @howardjohn!

/lgtm
/approve

**Comment by robscott** [comment]
I think we've got enough other LGTMs on this one to remove the hold.

/hold cancel

**Review by sunjayBhatia — APPROVED** [review]
/lgtm

**Comment by howardjohn** [comment]
The grpc implementation is incorrect here - the limit was never raised.

**Pull request merged**
